### PR TITLE
Preliminary NodeJS support

### PR DESCRIPTION
**Pull request by Jongy**
## Description
Small PoC that modifies our `perf` execution to add `perf inject --jit` - making it automatically handle processes that support `perf`'s "JIT" protocol, such as `node --perf-prof`.

This is less nice because it requires modifying code / configuration, unlike other profilers which require no changes at all (and besides async-profiler, truly non-intrusive to the app).

But... it just works :)

TODOs:
* [x] Add better explanation in README
* [ ] Add coloring for Node frames / builtins (perhaps color all frames that come from `/tmp/jitted-*.so`?)
* [x] Clean up jitted leftovers? After `perf inject --jit` or when gProfiler goes down?
* [x] Run `perf inject --jit` only if needed (to avoid overhead); or prove its overhead when there are no `jitdump` map entries is null
* [ ] Think about making this automatic (i.e, enabling `--perf-prof` in runtime)
* [x] ~~Possibly `-k 1` is not supported on all systems; in which case, don't use it and don't run `perf inject --jit`. Test it on some 2.6 kernel~~ now it's not automatically selected, so I don't mind
* [x] Add sanity tests

## Motivation and Context
NodeJS profiling support :)

## How Has This Been Tested?
Local sanity on `docker run -it --rm node:14 --perf-prof --interpreted-frames-native-stack` which runs `while (1) { }`

## Screenshots
Screenshot from 2021-06-28 01-15-52

## Checklist:
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have updated the relevant documentation.
- [ ] I have added tests for new logic.

### CODE REVIEW

**Comment by Jongy** [comment]
Since (for now) this ain't automatic and requires the user to pass `--perf-prof` in their app, I decided to also require the user to pass `--nodejs-mode perf` in order to enable the `perf inject --jit` part (avoiding it when not needed, for performance reasons)

**Comment by Jongy** [comment]
@michelhe , *now* it will pass

**Comment by Jongy** [comment]
Merged https://github.com/Granulate/gprofiler/pull/113 into here

**Comment by michelhe** [comment]
Tests still failing

**Comment by Jongy** [comment]
Kill me why, but doesn't work when Node runs on the host. Well, the `inject --jit` part does work:
```
node;__libc_start_main;node::Start;node::NodeMainInstance::Run;node::LoadEnvironment;node::StartExecution;node::StartExecution;node::ExecuteBootstrapper;v8::Function::Call;v8::internal::Execution::Call;v8::internal::(anonymous namespace)::Invoke;Builtin:JSEntry;Builtin:JSEntryTrampoline;InterpretedFunction: internal/main/run_main_module.js:1;InterpretedFunction:executeUserEntryPoint internal/modules/run_main.js:69;InterpretedFunction:Module._load internal/modules/cjs/loader.js:725;InterpretedFunction:Module.load internal/modules/cjs/loader.js:938;InterpretedFunction:Module._extensions..js internal/modules/cjs/loader.js:1094;InterpretedFunction:Module._compile internal/modules/cjs/loader.js:1039;[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];[/tmp/perf-91190.map];asm_sysvec_hyperv_stimer0_[k];sysvec_hyperv_stimer0_[k];irq_exit_rcu_[k];do_softirq_own_stack_[k];asm_call_sysvec_on_stack_[k];__softirqentry_text_start_[k]
``` 

but our fibonacci functions remains an unknown entry.

Possibly a difference in versions between the Node installed on the host vs the one we run in container...

**Comment by Jongy** [comment]
> Possibly a difference in versions between the Node installed on the host vs the one we run in container...

Pfft, seems like this was the problem. I'm not even sure which version did it have installed. I'll test other LTS versions (12.x, 14.x) and post here; possibly will also add them as tests

**Comment by Jongy** [comment]
> > Possibly a difference in versions between the Node installed on the host vs the one we run in container...
> 
> Pfft, seems like this was the problem. I'm not even sure which version did it have installed. I'll test other LTS versions (12.x, 14.x) and post here; possibly will also add them as tests

Ran the tests Dockerfile with `node:12` and `node:14`. They both exhibit the same results as https://github.com/Granulate/gprofiler/pull/126#issuecomment-882128624. I don't understand Node's "jitdump policy" (why does my fibonacci has no symbol in that case ....?) but that's out of scope for the PR, I guess.

@michelhe can review

File: gprofiler/main.py [hunk 1]
```diff
@@ -551,7 +563,10 @@ def parse_cmd_args():
         parser.error("--perf-dwarf-stack-size maximum size is 65528")
 
     if args.perf_mode in ("dwarf", "smart") and args.frequency > 100:
-        parser.error("--profiling-frequency|-f can't be larger than 100 when using --perf-mode smart|dwarf")
+        parser.error("--profiling-frequency|-f can't be larger than 100 when using --perf-mode 'smart' or 'dwarf'")
```
Comment (michelhe) [inline gprofiler/main.py:566]: Hmm first time I notice this.
I don't think its a good idea to deny users the option to sample at a higher frequency. We could instead warn about disk space implications.

File: gprofiler/perf.py [hunk 2]
```diff
@@ -18,21 +20,35 @@
 logger = get_logger_adapter(__name__)
 
 
+@functools.lru_cache(maxsize=1)
+def perf_path() -> str:
+    return resource_path("perf")
```
Comment (michelhe) [inline gprofiler/perf.py:25]: why not do the lru_cache on `resource_path` itself?

File: gprofiler/perf.py [hunk 3]
```diff
@@ -18,21 +20,35 @@
...
+        inject_jit: bool,
+        extra_args: List[str],
+    ):
         self._frequency = frequency
         self._stop_event = stop_event
         self._output_path = output_path
-        self._extra_args = extra_args
         self._type = "dwarf" if is_dwarf else "fp"
+        self._inject_jit = inject_jit
+        self._extra_args = extra_args + (["-k", "1"] if self._inject_jit else [])
```
Comment (michelhe) [inline gprofiler/perf.py:46]: why is it needed for inject-jit ?

File: gprofiler/perf.py (hunk 2 above)
Comment (Jongy) [reply gprofiler/perf.py:25]: Smart one

File: gprofiler/perf.py [hunk 4]
```diff
@@ -76,8 +92,16 @@ def wait_and_script(self) -> str:
         assert self._process is not None
         logger.debug(f"perf stderr: {self._process.stderr.read1(4096)}")
 
+        if self._inject_jit:
+            inject_data = Path(f"{str(perf_data)}.inject")
+            run_process(
+                [perf_path(), "inject", "--jit", "-o", str(inject_data), "-i", str(perf_data)],
```
Comment (michelhe) [inline gprofiler/perf.py:98]: I guess we can't pass same file for `-i` and `-o` right ?

File: gprofiler/perf.py (hunk 3 above)
Comment (Jongy) [reply gprofiler/perf.py:46]: It needs to correlate timestamps in the `jitdump` file with those from `perf.data`. So both need to use the same clock

File: gprofiler/main.py (hunk 1 above)
Comment (Jongy) [reply gprofiler/main.py:566]: It's not about disk space but about CPU usage. Anyway, that limit is indeed less relevant, we don't want to deny a conscious user from doing whatever they want. Will remove in another PR

File: tests/conftest.py [hunk 5]
```diff
@@ -82,6 +88,12 @@ def gprofiler_exe(request, tmp_path: Path) -> Path:
     return tmp_path / "gprofiler"
 
 
+def _print_process_output(popen: subprocess.Popen) -> None:
```
Comment (michelhe) [inline tests/conftest.py:91]: if the process is hanging, this will hang as well.
add a timeout

File: gprofiler/perf.py (hunk 4 above)
Comment (Jongy) [reply gprofiler/perf.py:98]: Nope, it reads as a stream

File: tests/conftest.py [hunk 6]
```diff
@@ -93,20 +105,31 @@ def lower_privs():
...
+
+        # ensure, again, that it still alive (if it exited prematurely it might provide bad data for the tests)
+        try:
+            popen.wait(0)
+        except subprocess.TimeoutExpired:
+            pass
+        else:
+            _print_process_output(popen)
+            raise Exception(f"Command {command_line} exited unexpectedly during the test with {popen.returncode}")
+
```
Comment (michelhe) [inline tests/conftest.py:130]: There are no tests that actively kill the process?

File: tests/conftest.py (hunk 5 above)
Comment (Jongy) [reply tests/conftest.py:91]: It's called only after kill/after it has exited

File: tests/test_sanity.py [hunk 7]
```diff
@@ -67,6 +68,19 @@ def test_rbspy(
...
 
+@pytest.mark.parametrize("runtime", ["nodejs"])
+def test_nodejs(
+    tmp_path: Path,
+    application_pid: int,
+    assert_collapsed,
+    gprofiler_docker_image: Image,
+    runtime: str,
+) -> None:
+    with SystemProfiler(1000, 3, Event(), str(tmp_path), "fp", True, 0) as profiler:
```
Comment (michelhe) [inline tests/test_sanity.py:79]: I genuinely don't like the invocation here, too many const arguments. kwargs will look better

**Review by michelhe — COMMENTED** [review]
LGTM, notify me when ready for merge

File: tests/test_sanity.py (hunk 7 above)
Comment (Jongy) [reply tests/test_sanity.py:79]: k

File: tests/conftest.py (hunk 6 above)
Comment (Jongy) [reply tests/conftest.py:130]: Nope, why would we (it's the process being profiled, and these tests check just the profiling of it)

File: gprofiler/utils.py [hunk 8]
```diff
@@ -46,6 +46,7 @@
 hostname: Optional[str] = None
 
 
+@lru_cache(maxsize=None)
```
Comment (michelhe) [inline gprofiler/utils.py:49]: use `functools.cache` if you are already passing None.

image

Comment (Jongy) [reply gprofiler/utils.py:49]: Too new (missing in e.g 3.7)

**Review by michelhe — DISMISSED** [review]

**Review by michelhe — APPROVED** [review]

**Pull request merged**